ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_device_op.hpp

```diff
@@ -0,0 +1,83 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <optional>
+#include <variant>
+
+#include "ttnn/tensor/tensor.hpp"
+#include "ttnn/core.hpp"
+#include "ttnn/device_operation.hpp"
+#include "ttnn/types.hpp"
+#include "ttnn/operations/conv2d/conv2d.hpp"
+#include "ttnn/deprecated/tt_dnn/op_library/sliding_window_op_infra/sliding_window.hpp"
+
+
+namespace ttnn::operations {
+namespace pool {
+
+inline uint32_t ceil_multiple_of(uint32_t n, uint32_t m) {
+    return (uint32_t) std::ceil((float) n / m) * m;
+}
+
+// new maxpool uop -- called from the macro-op
+struct MaxPoolNew {
+    struct operation_attributes_t {
+        tt::tt_metal::SlidingWindowConfig sliding_window_config_;
+        DataType output_dtype_;
+        MemoryConfig memory_config_;
+    };
+
+    struct tensor_args_t {
+        const Tensor& input_tensor_;
+    };
+
+    using shape_return_value_t = ttnn::Shape;
+    using tensor_return_value_t = Tensor;
+
+    struct MultiCore {
+        struct shared_variables_t {
+            KernelHandle reader0_kernel;
+            KernelHandle reader1_kernel;
+            CBHandle raw_in_cb;
+            CBHandle cb_out;
+            uint32_t ncores;
+            uint32_t ncores_w;
+        };
+
+        using cached_program_t = ttnn::device_operation::CachedProgram<shared_variables_t>;
+
+        static cached_program_t create(const operation_attributes_t& operation_attributes,
+                                       const tensor_args_t& tensor_args,
+                                       tensor_return_value_t& output_tensor);
+        static void override_runtime_arguments(cached_program_t& cached_program,
+                                               const operation_attributes_t& operation_attributes,
+                                               const tensor_args_t& tensor_args,
+                                               tensor_return_value_t& output_tensor);
+        static cached_program_t max_pool_2d_multi_core_sharded_with_halo_v2_new(const Tensor &input,
```

Comment:
These are defined in device_operation header, but not implemented in device_operation.cpp
My preference is to keep declaration and definition aligned between files unless absolutely necessary to do otherwise.

@arakhmati , as far as I understand this is based on the example?

Reply:
I expect this to be declared in `MultiCoreMaxPoolProgramFactory.hpp`, implemented in `MultiCoreMaxPoolProgramFactory.cpp` and if we want to avoid including that header here - it can be forward declared.

My thinking here is next:
There is no need for consumers of MaxPool2D Device Operation to be aware about these programs. 
They pretty much a detail of implementation of the device operation itself.

Reply:
I got rid of this function, so no longer here :)